yt_dlp/utils.py

```diff
@@ -5284,107 +5285,174 @@ def load_plugins(name, suffix, namespace):
+                            If `True`, any non-compatible object will first be
+                            converted into a string and then traversed into.
+
+    @return                 The result of the object traversal.
+                            If successful, `get_all=True` and the path branches at least once
+                            a list of results is returned instead.
+                            If unsuccessful or the result is `None` it will return `default`.
+    """
     if not casesense:
-        _lower = lambda k: (k.lower() if isinstance(k, str) else k)
-        path_list = (map(_lower, variadic(path)) for path in path_list)
+        _casefold = lambda k: k.casefold() if isinstance(k, str) else k
+
+    if isinstance(expected_type, type):
+        type_test = lambda val: val if isinstance(val, expected_type) else None
+    elif expected_type is not None:
+        type_test = lambda val: try_call(expected_type, args=(val,))
+    else:
+        type_test = IDENTITY
+
+    is_sequence = lambda x: isinstance(x, collections.abc.Sequence) and not isinstance(x, str)
+
+    def _traverse_obj(start_obj, path):
+        has_branched, results = _traverse_path(start_obj, path)
+        if results:
+            if get_all and has_branched:
+                return list(map(type_test, results))
+            else:
+                return type_test(results[0])
+        return None
 
-    def _traverse_obj(obj, path, _current_depth=0):
-        nonlocal depth
+    def _traverse_path(start_obj, path):
         path = tuple(variadic(path))
-        for i, key in enumerate(path):
-            if None in (key, obj):
-                return obj
-            if isinstance(key, (list, tuple)):
-                obj = [_traverse_obj(obj, sub_key, _current_depth) for sub_key in key]
+        objs = [start_obj]
+        has_branched = False
+
+        result_objs = []
+        for key in path:
+            if not objs:
+                return has_branched, result_objs
+
+            if is_user_input and key == ":":
                 key = ...
 
-            if key is ...:
-                obj = (obj.values() if isinstance(obj, dict)
-                       else obj if isinstance(obj, (list, tuple, LazyList))
-                       else str(obj) if traverse_string else [])
-                _current_depth += 1
-                depth = max(depth, _current_depth)
-                return [_traverse_obj(inner_obj, path[i + 1:], _current_depth) for inner_obj in obj]
-            elif isinstance(key, dict):
-                obj = filter_dict({k: _traverse_obj(obj, v, _current_depth) for k, v in key.items()})
-            elif callable(key):
-                if isinstance(obj, (list, tuple, LazyList)):
-                    obj = enumerate(obj)
+            if not casesense and isinstance(key, str):
+                key = key.casefold()
+
+            result_objs.clear()
+            for obj in objs:
```

Comment:
I think this loop should be a function
```py
def get_key(objs, key):
    for obj in objs:
        ...
        yield result

objs = [v for v in get_key(objs, key) if v is not None]
```

Reply:
Extracting the loop out will be hard since we loose track of if we have branched or not. It can be done using a nested function inside with a local `has_branched` value though. I can try implementing that if desired. This would also make converting to a depth first search on `get_all=False` easier.